### ISSUE TITLE
Templates, filtering and image sizes

### ISSUE THREAD

**Issue by maehw** [xtech/self-o-mat#40]
While browsing the self-o-mat code and where I may want to place changes, I've developed some diagrams which describe the software architecture (or my understanding of it) as well as the image processing chain. This is what I came up with for a combination of https://github.com/xtech/self-o-mat/blob/af62ea58fea9765efc8652813146a01b50971c81/src/logic/ImageProcessor.cpp#L66 and https://github.com/xtech/self-o-mat/blob/af62ea58fea9765efc8652813146a01b50971c81/src/logic/ImageProcessor.cpp#L164

![image-processing-chain](https://github.com/user-attachments/assets/6712cb5d-08f9-4cf2-9b16-6794e79076ca)

I have some remarks/questions which I'd be happy if someone commented on (most likely you, @ClemensElflein ):

1. When using templates. Does the `applyFilter` function call really work as expected? https://github.com/xtech/self-o-mat/blob/af62ea58fea9765efc8652813146a01b50971c81/src/logic/ImageProcessor.cpp#L129 -- Shouldn't variables `result` or `decodedImageROI` be used to apply the filter? In my understanding `decodedImage` won't be used anymore. Can you confirm? (Maybe it just did not pop up because the default filter is not very prominent.)
2. Why did you call it "compression" here? I think it's "just" applying the alpha channel of the template... or blending template and resized image together... - nothing really "compressed" here, right? https://github.com/xtech/self-o-mat/blob/af62ea58fea9765efc8652813146a01b50971c81/src/logic/ImageProcessor.cpp#L153
3. When no template is applied, the target size is 1773 pixels x 1182 pixels (hard-coded) -- that's a perfect aspect ratio of 1.5 (or 3:2). Assuming 300 dpi printers, this yields 5.91" x 3.94"print size. However, [we've documented](https://github.com/xtech/self-o-mat/wiki/Hardware) that the "Hagaki" format ("postcard") has 148 mm x 100 mm or 5.83" x 3.94". Do you see a problem in the .2mm difference? Not too many pixels lost probably.. but do you think this will cause the printer to re-scale unnecessarily?
4. I've never used templates before, but I'd like to give them a try. In https://github.com/xtech/self-o-mat/wiki/Configuration#templates we started documenting them in the wiki. How are the other hard-coded values related? https://github.com/xtech/self-o-mat/blob/af62ea58fea9765efc8652813146a01b50971c81/src/logic/ImageProcessor.cpp#L273-L274 -- Yet another slightly different aspect ratio (approx. 1.518). 1864 pixels x 1228 pixels seems to be "the perfect size" and would match [the template file you provided once](https://user-images.githubusercontent.com/6305922/143136175-b8ac1162-d455-4087-bf12-62776510a863.png), but why this size?

I'd be happy if we could make sense out of these numbers.

**Comment by ClemensElflein** [comment]
It has been some time, so I might be off about some of the things, so forgive me if anything doesn't seem to make sense. The drawing looks good to me.

1. Yes looks like you are right - this would mean that filtering in combination with the Template is not working as expected
2. Good question, this indeed does not do any compression at all. It combines template and image, must be an old log or copy/paste mistake.
3. I'm not sure where this number came from, sorry. Is it really 0.2mm difference or is it 2mm difference? In my head the most common picture format is 10x15 cm, so if it's 2mm difference I probably did 10x15 cm format (now that I'm reading up on it, it doesn't seem to be _exactly_ 15x10 cm after all, but it would explain why those numbers are there).
4. If I remember correctly, these pixels are the screen resolution of the provided LCD of the self-o-mat boxes. The scaling was done in a way, that the template matches the screen's width or height (or both, if aspect ratio was correct).

Hope any of that helps

**Comment by ClemensElflein** [comment]
You are right, it has the screen one separately. So I think there's really no reason for the different resolution
